[inductor] Fix logging for run_and_get_cpp_code

masnesral opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128794

Summary: Found during testing with remote caching: Use the same output logger object between graph.py and codecache.py since it's patched in `run_and_get_cpp_code`. That allows us to capture any logging produced from the codecache path when using `run_and_get_cpp_code`. I'm also fixing a few tests that were passing mistakenly because logging was missing.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D58796516](https://our.internmc.facebook.com/intern/diff/D58796516)

masnesral committed:
[inductor] A couple minor changes to fix inductor tests affected by remote caching

Summary:
1) Change run_and_get_cpp_code() to use the same mechanism used by the other run_and_get_ methods to capture generated code.
2) Disable remote caching deliberately on one test that installs a patch and expects it to be called.

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] A couple minor changes to fix inductor tests affected by remote caching"


Summary:
1) Change run_and_get_cpp_code() to use the same mechanism used by the other run_and_get_ methods to capture generated code.
2) Disable remote caching deliberately on one test that installs a patch and expects it to be called.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] A couple minor changes to fix inductor tests affected by remote caching"


Summary:
1) Disable remote caching deliberately on one test that installs a patch and expects it to be called.
2) Use the same output logger object between graph.py and codecache.py since it's patched in `run_and_get_cpp_code`

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Fix logging for run_and_get_cpp_code"


Summary: Found during testing with remote caching: Use the same output logger object between graph.py and codecache.py since it's patched in `run_and_get_cpp_code`. That allows us to capture any logging produced from the codecache path when using `run_and_get_cpp_code`. I'm also fixing a few tests that were passing mistakenly because logging was missing.

[ghstack-poisoned]

masnesral commented on test/inductor/test_cpu_cpp_wrapper.py:
```diff
@@ -255,7 +255,8 @@ class BaseTest(NamedTuple):
         BaseTest(
             "test_multi_threading",
-            code_string_count={"py::gil_scoped_release release;": 1},
+            # Two threads compile, so we expect the output code to be printed twice.
```
We weren't capturing the output from the cache hit path so I think this was passing by mistake

masnesral commented on test/inductor/test_cpu_repro.py:
```diff
@@ -1933,10 +1933,8 @@ def _internal_check(
-            # which fails to change thread number with native thread pool
             with set_num_threads(1):
-                _internal_check(fn, inps, _target_code_check_not="aten.scatter_reduce_")
+                _internal_check(fn, inps, "aten.scatter_reduce_")
```
The issue above is closed and I think removing this was an oversight because the test was passing mistakenly. I believe it was passing before because the code output was empty (because we weren't capturing it) so the scatter_reduce string was not present :) . But after this change, the code output is populated and the string is present and this test is now correct. @leslie-fang-intel what do you think?

leslie-fang-intel replied:
oh, thanks for the fix of `run_and_get_cpp_code` @masnesral. However, this line of change feels wrong to me, since when running with single thread, we expect the `aten.scatter` will go into the cpp backend codegen instead fallback to  `aten.scatter_reduce_`. That's why we check `aten.scatter_reduce_` not in generated code intentionally.

leslie-fang-intel replied:
BTW: if we run a function several times but with different config in a single UT, should we disable the graph cache you mentioned?

masnesral replied:
Oh, ok. Well this test was passing by mistake then. I can fix it. 

> BTW: if we run a function several times but with different config in a single UT, should we disable the graph cache you mentioned?

Well any inductor configuration that we think could affect its compilation output is included in the key calculation for the cache. So if configuration changes, then we won't serve a cache entry created with another configuration. But I was not aware of `torch.get_num_threads()` and I really wouldn't have expected that it could affect compilation. That is not currently included in the cache key. Do you think it should? Would we want to avoid serving a cache entry that was generated with more than a single thread??

leslie-fang-intel replied:
Despite of the fallback or codegen for `aten.scatter`, I think the change of the number threads will affect the CPP backend compilation in a more broad way, as in https://github.com/pytorch/pytorch/blob/2f88597aad145f6b37d0208bc58f087510e14565/torch/_inductor/codegen/cpp.py#L1890, the parallel depth will changes along with the number of threads. So, I guess we need to add the threads number into the graph cache key. cc @jgong5 for more comments.

masnesral replied:
Ok. I think if we were to include the number of threads directly in the key calculation, that wouldn't be great for cache performance because, for example, we'd have different entries for identical graphs that were compiled in an environment with 44 vs. 42 threads. But we could differentiate 1 thread vs. many and include that boolean in the key calculation. @eellison what do you think about this?

masnesral committed:
Update on "[inductor] Fix logging for run_and_get_cpp_code"


Summary: Found during testing with remote caching: Use the same output logger object between graph.py and codecache.py since it's patched in `run_and_get_cpp_code`. That allows us to capture any logging produced from the codecache path when using `run_and_get_cpp_code`. I'm also fixing a few tests that were passing mistakenly because logging was missing.

[ghstack-poisoned]

masnesral commented on test/inductor/test_cpu_repro.py:
```diff
@@ -1933,10 +1935,16 @@ def _internal_check(
-            # which fails to change thread number with native thread pool
             with set_num_threads(1):
-                _internal_check(fn, inps, _target_code_check_not="aten.scatter_reduce_")
+                # When running with a single thread, we expect the aten.scatter will go
```
@leslie-fang-intel

leslie-fang-intel replied:
Thanks, this change looks good to me.

masnesral commented:
@pytorchbot merge

oulgen commented:
@oulgen has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D58796516).
